### PR TITLE
Long running query causes DTU spikes to 100%

### LINKED ISSUE #2279: Long running query causes DTU spikes to 100%
## Problem

Purge Log Buffer scheduler is failing when EventLog table gets big. Long running query causes DTU spikes on a server and finishes with a timeout exception.

## Rootcause

Scheduler is stopped for a long time or Purge Log process is not activated: both are possible scenarios when it can happen. It makes EventLog table overfilled (up to 1M and more) and causes exceptions on a next scheduler run.

## Solution

Based on execution plan and numerous testings, DELETE command takes the longest part. It is not effective to run DELETE on a huge amount of records by once.
The solution proposed is to run purge process with a smaller batches. The size of batches is defined in a host settings and under administrator's control. Such limitation makes the process stable and highly loaded table gets cleared well in a time.

ESW ticket DNN-20358

### PR DESCRIPTION
**Pull request by mikebigun**
fixes #2279

### CODE REVIEW

File: Website/Providers/DataProviders/SqlDataProvider/09.03.00.SqlDataProvider [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+	DECLARE @PurgeBatchCountSettingName NVARCHAR(50) = N'PurgeEventLogBatchCount'
+	SELECT @PurgeBatchCount = 
+		CASE WHEN ISNUMERIC(SettingValue) > 0 
+			THEN CAST(SettingValue AS INT)
+			ELSE @PurgeBatchCount 
+		END
+	FROM {databaseOwner}[{objectQualifier}HostSettings]
+	WHERE SettingName = @PurgeBatchCountSettingName
+
+	;WITH logcounts AS
```
[the remark this answers is not in the document]
Comment (sleupold) [reply Website/Providers/DataProviders/SqlDataProvider/09.03.00.SqlDataProvider]: I suggest using this stored procedure:
CREATE PROCEDURE {databaseOwner}[{objectQualifier}PurgeEventLog]
AS
BEGIN
	DECLARE @keepDays        INT = IsNull((SELECT SettingValue 
									        FROM  {databaseOwner}[{objectQualifier}HostSettings] 
									        WHERE SettingName = N'EventLogKeepForDays'
									        AND IsNumeric(SettingValue)), -1);
    DECLARE @PurgeBatchCount INT = IsNull((SELECT CAST(SettingValue AS INT)
                                        	FROM {databaseOwner}[{objectQualifier}HostSettings] 
											WHERE SettingName = N'PurgeEventLogBatchCount'
											AND IsNumeric(SettingValue) > 0 ), 10000);

    -- delete entries, older than specified time:
	IF IsNull(@keepDays, -1) >= 0
		DELETE FROM L
		FROM  {databaseOwner}[{objectQualifier}EventLog]       L 
		JOIN  {databaseOwner}[{objectQualifier}EventLogConfig] C ON C.ID = L.LogConfigID
		WHERE (C.KeepMostRecent < 0 AND LogCreateDate < DateAdd(d, @keepDays, GetDate()));

    -- delete entries, older than number to keep (as specified for log type):
	DELETE FROM {databaseOwner}[{objectQualifier}EventLog]
	 FROM
	( SELECT TOP @PurgeBatchCount 
	         L.LogEventID,
             C.KeepMostRecent,
             ROW_NUMBER() OVER (PARTITION BY L.LogConfigID ORDER BY L.LogCreateDate DESC) SeqNumber
       FROM  {databaseOwner}[{objectQualifier}EventLog]       L
       JOIN  {databaseOwner}[{objectQualifier}EventLogConfig] C ON C.ID = L.LogConfigID
       WHERE C.KeepMostRecent <> -1 OR C.LoggingIsActive =  0
    ) AS S
	 WHERE SeqNum

**Comment by bdukes** [comment]
@ohine am I remembering correctly that you had implemented a batched event log deletion at some point, or was that a different issue?

File: Website/Providers/DataProviders/SqlDataProvider/09.03.00.SqlDataProvider (hunk 1 above)
[the remark this answers is not in the document]
Comment (mikebigun) [reply Website/Providers/DataProviders/SqlDataProvider/09.03.00.SqlDataProvider]: @sleupold 

I'm a bit confused.
I have checked all scripts starting from initial one to the latest version and did not found EventLogKeepForDays setting. I also checked all existing ones and do not see any of possible setting that could work as you described.  May I ask you to point me on that setting in the scripts?
I did a minimal modification to existing SP. Nothing was deleted.

What I can easily do is:
- remove variables from constants
- replace CTE with JOIN (will do additional testing)

For 'EventLogKeepForDays' I need additional info. Also, if this part does not exist and should be implemented, I would go with a separate issue. 

cc: @tpluscode

[the remark this answers is not in the document]
Comment (sleupold) [reply Website/Providers/DataProviders/SqlDataProvider/09.03.00.SqlDataProvider]: @mikebigun 
please check out https://dnntracker.atlassian.net/browse/DNN-8668
I was of the impression, it got merged into DNN 8

**Comment by bdukes** [comment]
Ah, that was schedule history, rather than event log

**Comment by mikebigun** [comment]
@sleupold @tpluscode 

To summarize above:

1. We do not have 'EventLogKeepForDays' implemented. Related [PR](https://github.com/dnnsoftware/Dnn.Platform/pull/1503) is closed and not merged.
2. CTE does not use temp tables or memory. Based on msdn it is temporary named derived table. In other words, it works similarly to subquery in terms of performance, but has many other advantages. The main we use here is better readability. Moreover, I have performed additional testing CTE vs. subquery. I got same execution plans and execution time (tested on 1M records). To my opinion there is no objective reasons to replace CTE.
3. Constants and variables will be fixed as suggested above.

**Comment by sleupold** [comment]
@mikebigun 
1. ok,  I already included it into my TurboScripts - it is just this single statement needed to function (and the host setting), allowing to constrain EventLog to a max age (just if setting is present, otherwise there is no relevant performance decrease for this sproc).
2. interesting, I was comparing query plans of join and CTE yesterday and noticed some differences. Which version of SQL Server and which database compatibility level are you using? The new query plan analyzer in SQL 2014 might handle it differently than the old one.
3. fine, thanks!

**Comment by mikebigun** [comment]
@sleupold Ok.

For the point 2: The query (without CTE) I tested is following:

```
DELETE  lc
FROM (SELECT TOP(@PurgeBatchCount)
		LogConfigID, 
		ROW_NUMBER() OVER(PARTITION BY LogConfigID ORDER BY LogCreateDate DESC) AS 
          logEventSequence	
	  FROM {databaseOwner}[{objectQualifier}EventLog]) as lc
INNER JOIN {databaseOwner}[{objectQualifier}EventLogConfig] elc ON elc.ID = lc.LogConfigID
	WHERE elc.KeepMostRecent <> -1
		AND lc.logEventSequence > elc.KeepMostRecent
```

It is different a bit versus that one you proposed above. In fact, I'm avoiding additional JOIN. I do deletion directly from subquery and not from EventLogs. I do not need to match subquery with a parent EventLogs. So then execution plan of my query and yours is different.
In your plan I see one more JOIN.

But the point is that CTE does not impacting query. Above query works absolutely similarly to CTE.

I'm using: Microsoft SQL Server 2012 - 11.0.2100.60 (X64)

**Comment by sleupold** [comment]
I ran the following 4 statements and the second seems to perform best on a small data set - unfortunately I don't have the time to create a proper test bed:
`	DELETE  TOP (1000) FROM dbo.[EventLog] 
	 FROM
	( SELECT 
	         L.LogEventID,
             C.KeepMostRecent,
             ROW_NUMBER() OVER (PARTITION BY L.LogConfigID ORDER BY L.LogCreateDate DESC) SeqNumber
       FROM  dbo.[EventLog]       L
       JOIN  dbo.[EventLogConfig] C ON C.ID = L.LogConfigID
       WHERE C.KeepMostRecent <> -1 OR C.LoggingIsActive =  0
    ) AS S
	 WHERE SeqNumber > KeepMostRecent;
GO
	DELETE  FROM dbo.[EventLog]
	 FROM
	( SELECT TOP (1000)
	         L.LogEventID,
             C.KeepMostRecent,
             ROW_NUMBER() OVER (PARTITION BY L.LogConfigID ORDER BY L.LogCreateDate DESC) SeqNumber
       FROM  dbo.[EventLog]       L
       JOIN  dbo.[EventLogConfig] C ON C.ID = L.LogConfigID
       WHERE C.KeepMostRecent <> -1 OR C.LoggingIsActive =  0
    ) AS S
	 WHERE SeqNumber > KeepMostRecent;
GO
	DELETE TOP (1000) FROM L
	FROM
	(SELECT 
	        LogEventID,
			LogConfigID,
            ROW_NUMBER() OVER (PARTITION BY LogConfigID ORDER BY LogCreateDate DESC) SeqNumber
	  FROM  dbo.[EventLog]) AS L
	  WHERE SeqNumber > (SELECT KeepMostRecent FROM dbo.[EventLogConfig]  C WHERE C.ID = L.LogConfigID);
GO
	DELETE FROM L
	FROM
	(SELECT TOP (1000) 
	        LogEventID,
			LogConfigID,
            ROW_NUMBER() OVER (PARTITION BY LogConfigID ORDER BY LogCreateDate DESC) SeqNumber
	  FROM  dbo.[EventLog]) AS L
	  WHERE SeqNumber > (SELECT KeepMostRecent FROM dbo.[EventLogConfig]  C WHERE C.ID = L.LogConfigID);
GO`

**Comment by mikebigun** [comment]
@sleupold 

> I am not sure your query is correct, because your purgeBatchCount is applied prior to the Where condition. 

This is our goal. We do not want to apply 'delete' along with 'where' to the whole table. I want to take a subset of rows and determine whether there is something to delete. Sequence number ordered by date is what I need. That's it.

> Without an "ORDER BY", SELECT TOP is not deterministic and could return just the latest records, which might be filtered afterwards and result in no records being filtered at all.

I totally disagree. Every subset of N rows ranked by sequence number from newest to oldest log (that works to any subset, no matter latest rows, random rows or any other). If 'select top' returns latest rows, **nothing will be removed if seq number is less than KeepMostRecent** for the particular log type. Hence, I do not really care about deterministic. At the end you will get subset when nothing to delete. My protection mechanism to not delete newest entries in subset is the same as for the whole table. 

Secondly, query 1 and 2 you provided are incorrect. They removes everything from EventLog. They don't have matching to a subset. I assume you did a typo. Seems you wanted to make a join of parent to subquery by event id. I corrected it in a test below.

To finally close this discussion, I would like to share this [doc](https://docs.google.com/document/d/1KBRwvI7uWSzY_rmw1SmAAaQkU-FHpDXw71P1UWAPRRM/edit)

Let me please know the number of a query I should commit. I will do it in my next commit so we will close this PR.

File: Website/Providers/DataProviders/SqlDataProvider/09.03.00.SqlDataProvider [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+/*****                                                  *****/
+/*****                                                  *****/
+/***** Note: To manually execute this script you must   *****/
+/*****       perform a search and replace operation     *****/
+/*****       for {databaseOwner} and {objectQualifier}  *****/
+/*****                                                  *****/
+/************************************************************/
+
+IF EXISTS (SELECT * FROM dbo.sysobjects
+	WHERE id = object_id(N'{databaseOwner}[{objectQualifier}HostSettings]') and OBJECTPROPERTY(id, N'IsUserTable') = 1)
```
Comment (mitchelsellers) [inline Website/Providers/DataProviders/SqlDataProvider/09.03.00.SqlDataProvider]: This check seems un-necessary as if this table isn't there it is impossible for this process to work

**Comment by mikebigun** [comment]
@tpluscode Can you check now?

**Comment by sleupold** [comment]
@mikebigun 
sorry for the late reply, I was too busy to checkout your reply and document. 
Reviewing the queries you tested, IDs 1, 2. 3 and 4 should always delete the proper records, but query with ID 5 (which is included in the PR) is not guaranteed to delete records, depending on indexes and settings for keepMostRecent and PurgeEventLogBatchCount:
If the records in your TVE are ordered first by LogConfigID and second by LogCreateDate and the keepMostRecent for the first EventLog type  is null or larger than PurgeEventLogBatchCount, the WHERE condition might not match any records and no records are returned.

I agree, this is an edge case, but IMO one that should be avoided. I suggest using one of the other approaches, according to your tests, ID 3 or 4 would perform best.

**Comment by sleupold** [comment]
@mikebigun 
ouch, I just noticed you  missed to take care of event log types set to keep all records.
